kernel/aster-nix/src/process/sync/condvar.rs

```diff
@@ -0,0 +1,195 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use alloc::sync::Arc;
+use core::time::Duration;
+
+use aster_frame::sync::{guard_lock, MutexGuard, SpinLock};
+
+use super::Pauser;
+use crate::Errno;
+
+/// The lock error type for situations where the condvar is not notified.
+pub enum LockErr<Guard> {
+    Timeout(Guard),
+    Interrupted(Guard),
+    Unknown(Guard),
+}
+/// LockResult, different from Rust std.
+/// The result of a lock operation.
+pub type LockResult<Guard> = Result<Guard, LockErr<Guard>>;
+
+impl<Guard> LockErr<Guard> {
+    pub fn into_guard(self) -> Guard {
+        match self {
+            LockErr::Timeout(guard) => guard,
+            LockErr::Interrupted(guard) => guard,
+            LockErr::Unknown(guard) => guard,
+        }
+    }
+}
+
+type WaitCounter = u64;
+type NotifyCounter = u64;
+
+/// A condition variable.
+/// pauser: The pauser object used to pause and resume threads.
+/// counter: The counter (WaitCounter, NotifyCounter).
+pub struct Condvar {
+    pauser: Arc<Pauser>,
+    counter: SpinLock<(WaitCounter, NotifyCounter)>,
+}
+
+impl Condvar {
+    /// Creates a new condition variable.
+    pub fn new() -> Self {
+        Condvar {
+            pauser: Pauser::new(),
+            counter: SpinLock::new((0, 0)),
+        }
+    }
+
+    /// Waits for the condition variable to be signaled or broadcasted.
+    pub fn wait<'a, T>(&self, guard: MutexGuard<'a, T>) -> LockResult<MutexGuard<'a, T>> {
```

Comment:
The `wait` method must provide the _atomicity_ property, which is described in Rust std's own words as follows:

> This function will atomically unlock the mutex specified (represented by guard) and block the current thread.

The unlock and block must happen atomically. Otherwise, a thread that waits on a conditional variable may never be woken up.

Consider the following simple, but typical usage of `Condvar`. This example involves two concurrent threads, one is executing the function `thread_a`, the other doing `thread_b`.

```rust
fn thread_a(mutex: &Mutex<bool>, condvar: &Condvar) -> Result<()> {
    let mut guard = mutex.lock(); // Label A
    loop {
        if check_cond(&*guard) {
            return;
        }

        guard = condvar.wait(guard)?; // Label B
    }
}

fn thread_b(mutex: &Mutex<u32>, condvar: &Condvar) -> Result<()> {
    { // Label C
        let mut guard = mutex.lock();
        set_cond(&mutex);
    } // Label C'
    condvar.notify();
}

fn check_cond(val: &u32) -> bool {
    *val == 123;
}

fn set_cond(val: &mut u32) {
    *val = 123;
}
```

Regardless of how the two threads may execute, we can be sure that the function `thread_a` will eventually return. Here is the reasoning.

Thanks to the atomicity guarantee of `Condvar`, the atomic section between label C and C' can only happens before label A or after label B. If it happens before label A, then the condition has already been set to true, so the `check_cond` function must return true, and thus `thread_a` returns. If the atomic section happens after B, then the `condvar.notify()` in `thread_b`must be called after `condvar.wait(..)` in `thread_a`. So the `wait` method must return.

Unfortunately, the proposed implementation based on counters does not satisfy the atomicity property and thus may never wake up `thread_a`.

Reply:
Thank you for your comments. 

I have reconsidered the race condition. For the `wait` function, there will be no problem as long as the atomicity of the counter modification is ensured, that is, the mutex is held until the counter modification is completed.

The states of the example can be abstracted as the following:

- A1: Check the condition, modify the counter, and release the lock
- A2: wait queue wait
- B1: Modify conditions and release lock
- B2: Modify counter, notify

And we have the following function calls sequences:

- A1-B1-A2-B2
- A1-A2-B1-B2
- A1-B1-B2-A2
- B1-B2-A1-break
- B1-A1-break

Counter modification is atomic, and `wait` can be considered successful after the counter is modified.
Therefore, if `wait` happens, the lock is kept until the counter is added. And `notify` will wake it up. If `notify` is called first, `thread_a` will never wait.